packages/clerk-js/src/ui/elements/Select.tsx

```diff
@@ -0,0 +1,338 @@
+import React, { isValidElement, PropsWithChildren, useState } from 'react';
+
+import { usePopover, useSearchInput } from '../../ui/hooks';
+import { Button, Flex, Icon, Text } from '../customizables';
+import { Caret, MagnifyingGlass } from '../icons';
+import { animations, PropsOfComponent } from '../styledSystem';
+import { colors, createContextAndHook } from '../utils';
+import { InputWithIcon } from './InputWithIcon';
+
+type UsePopoverReturn = ReturnType<typeof usePopover>;
+type UseSearchInputReturn = ReturnType<typeof useSearchInput>;
+
+type OptionBuilder<O> = (option: O, index?: number, isSelected?: boolean) => JSX.Element;
+
+type SelectProps<O> = {
+  options: O[];
+  value: O;
+  onChange: (option: O) => void;
+  placeholder?: string;
+  comparator?: (term: string, option: O) => boolean;
+  noResultsMessage?: string;
+  optionBuilder?: OptionBuilder<O>;
+};
+
+type SelectState<O = any> = Pick<SelectProps<O>, 'placeholder' | 'comparator' | 'noResultsMessage'> & {
+  popoverCtx: UsePopoverReturn;
+  searchInputCtx: UseSearchInputReturn;
+  optionBuilder: OptionBuilder<O>;
+  selectedOption: O;
+  select: (option: O) => void;
+  selectedItemRef: React.RefObject<HTMLDivElement>;
+  onTriggerClick: () => void;
+};
+
+const [SelectStateCtx, useSelectState] = createContextAndHook<SelectState>('SelectState');
+
+const defaultOptionBuilder = <O,>(option: O, _index?: number, isSelected?: boolean) => {
+  return (
+    <Flex
+      sx={theme => ({
+        width: '100%',
+        paddingLeft: theme.space.$2,
+        paddingRight: theme.space.$2,
+        ...(isSelected && { backgroundColor: theme.colors.$blackAlpha200 }),
+        '&:hover': {
+          backgroundColor: theme.colors.$blackAlpha200,
+        },
+      })}
+    >
+      {['string', 'number'].includes(typeof option) && <Text>{option}</Text>}
```

Comment:
Do we need this check here? What else do you expect to receive here?

Reply:
This is the default builder. In order to use the default builder, the Options must be numbers or strings. We cannot render an object there. In that case, a builder will need to be provided.

Reply:
✔️